Tasks/Common/ios-signing-common/ios-signing-common.ts

```diff
@@ -602,8 +598,93 @@ export async function getP12CommonName(p12Path: string, p12Pwd: string) {
+    const privateKeyPassword = p12Pwd ? p12Pwd : generatePassword();
+    openssl.arg(['pkcs12', '-in', p12Path, '-nocerts', '-passin', 'pass:' + p12Pwd, '-passout', 'pass:' + privateKeyPassword]);
+
+    //we pipe through grep so we we don't log the private key to the console.
+    //even if it's encrypted, it's noise and could cause concern for some users.
+    const grepPath: string = tl.which('grep', true);
+    const grep: ToolRunner = tl.tool(grepPath);
+    grep.arg(['friendlyName']);
+    openssl.pipeExecOutputToTool(grep);
+
+    let privateKeyName: string;
+    openssl.on('stdout', function (data) {
+        if (data) {
+            // find the private key name
+            data = data.toString().trim();
+
+            const match = data.match(/friendlyName: (.*)/);
+            if (match && match[1]) {
+                privateKeyName = match[1].trim();
+            }
+        }
+    });
+
+    await openssl.exec();
+    tl.debug('P12 private key name = ' + privateKeyName);
+    if (!privateKeyName) {
+        throw new Error(tl.loc('P12PrivateKeyNameNotFound', p12Path));
+    }
+
+    return privateKeyName;
+}
+
+/**
+ * Set the partition_id ACL so codesign has permission to use the signing key.
+ */
+async function setKeyPartitionList(keychainPath: string, keychainPwd: string, privateKeyName: string) {
+    // security set-key-partition-list -S apple: -s -l <privateKeyName> -k <keychainPwd> <keychainPath>
+    // n.b. This command could update multiple keys (e.g. an expired signing key and a newer signing key.)
+
+    if (privateKeyName) {
+        tl.debug(`Setting the partition_id ACL for ${privateKeyName}`);
+
+        // "If you'd like to run /usr/bin/codesign with the key, "apple:" must be an element of the partition list." - security(1) man page.
+        // When you sign into your developer account in Xcode on a new machine, you get a private key with partition list "apple:". However
+        // "security import a.p12 -k login.keychain" results in the private key with partition list "apple-tool:".
+        const partitionList = 'apple:';
+
+        let setKeyCommand: ToolRunner = tl.tool(tl.which('security', true));
+        setKeyCommand.arg(['set-key-partition-list', '-S', partitionList, '-s', '-l', privateKeyName, '-k', keychainPwd, keychainPath]);
+
+        // Watch for "unknown command". set-key-partition-list was added in Sierra (macOS v10.12)
+        let unknownCommandErrorFound: boolean;
+        setKeyCommand.on('errline', (line: string) => {
+            if (!unknownCommandErrorFound && line.includes('security: unknown command')) {
+                unknownCommandErrorFound = true;
+            }
+        });
+
+        try {
+            await setKeyCommand.exec();
+        } catch (err) {
+            if (unknownCommandErrorFound) {
+                // If we're on an older OS, we don't need to run set-key-partition-list.
+                console.log(tl.loc('SetKeyPartitionListCommandNotFound'));
```

Comment:
If we call `tl.debug` instead of `console.log`, users who have sys.debug set to false will probably wonder if its okay that security output "command not found: set-key-partition-list".